Add preliminary support for SCTP

DolceTriade opened the pull request:
Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Thanks for contributing!

<!-- Description of change -->

Partially implements #5719 and implements M1 for #20490

```release-note
Add partial support for SCTP
```

pchaigno commented on bpf/lib/icmp6.h:
```diff
@@ -316,6 +316,7 @@ static __always_inline int __icmp6_send_time_exceeded(struct __ctx_buff *ctx,
 			return DROP_WRITE_ERROR;
 
 		break;
+	/* FIXME: Add IPPROTO_SCTP support */
```
We should create an issue for this instead of leaving a note in the code. I think it can be an issue for all STCP leftovers if there are several.

DolceTriade replied:
Should we continue to use https://github.com/cilium/cilium/issues/5719?

pchaigno replied:
I think we can but we'll have to update it.

joestringer replied:
^^ If you are going to continue pursuing this @DolceTriade, then maybe it would be worthwhile for you to create an issue for it so that you can edit / update the status in the issue description. We can then point this one to the new one you create.

DolceTriade replied:
Created https://github.com/cilium/cilium/issues/20490

pchaigno commented on pkg/policy/api/rule_validation_test.go:
```diff
@@ -95,6 +95,25 @@ func (s *PolicyAPITestSuite) TestL7RulesWithNonTCPProtocols(c *C) {
 	err = validPortRule.Sanitize()
 	c.Assert(err, IsNil, Commentf("Saw an error for a L7 rule with DNS rules. This should be allowed."))
 
+	validSctpRule := Rule{
```
Nit:
```suggestion
	validSCTPRule := Rule{
```

DolceTriade replied:
Will fix as part of the rebase.

pchaigno commented on pkg/policy/visibility.go:
```diff
@@ -83,6 +83,7 @@ func NewVisibilityPolicy(anno string) (*VisibilityPolicy, error) {
 		if u8Prot == u8proto.ANY {
 			protos = append(protos, u8proto.TCP)
 			protos = append(protos, u8proto.UDP)
+			protos = append(protos, u8proto.SCTP)
```
Did you test L7 policies with this pull request? I thought we had other dependencies on the L4 protocol.

DolceTriade replied:
Mainly what was in the tests. I didn't explicitly test any L7 policies myself since I thought they would be irrelevant to the changes. I did test L4 policies.

pchaigno replied:
Ok. We should list L7 as a limitation then.

DolceTriade replied:
Sorry, I'm not sure I understand. If L7 policies are being tested as part of `cilium connectivity test` and other tests that pass with this patch, is it still a limitation?

pchaigno replied:
But we don't support policies for L7 protocols over SCTP, do we? It just doesn't break existing L7 policies.

DolceTriade replied:
Ahh. Yes. I see what you mean. Will note that.

pchaigno commented on bpf/lib/csum.h:
```diff
@@ -41,7 +41,17 @@ static __always_inline void csum_l4_offset_and_flags(__u8 nexthdr,
+		 * crc32c checksum in eBPF and will likely require
+		 * additional kernel support.
+		 */
+		off->offset = 0;
```
As discussed on Slack, this significantly limits the SCTP support (no policies for pod-to-VIP, no KPR, no BPF masquerading, no egress gateway, etc.). Thus, can we put this feature behind a disabled-by-default flag and document those limitations?

DolceTriade replied:
Agreed. I was uncertain on how such a flag could be implemented and wanted to clarify.

pchaigno commented on bpf/lib/lb.h:
```diff
@@ -398,8 +398,12 @@ static __always_inline int reverse_map_l4_port(struct __ctx_buff *ctx, __u8 next
+				/* This will change the SCTP checksum, which we cannot fix right now.
+				 * This will likely need kernel changes before we can drop this.
+				 */
+				if (nexthdr == IPPROTO_SCTP) return CTX_ACT_DROP;
```
I think we should put the feature behind a flag rather than introduce runtime checks in the datapath code like this.

joestringer replied:
I had requested explicit drop for this case to ensure that we don't end up just passing packets on with malformed checksums. Could we just wrap this in the feature flag check? It's kinda helpful to have an explicit mention here as well to highlight that this path is broken in the SCTP support.

pchaigno replied:
Yeah, I'm happy with this if it's behind a disabled-by-default feature flag. I just don't want everyone to pay the performance & complexity penalty for a feature they may not use.

DolceTriade replied:
SGTM

pchaigno reviewed (changes requested):
Thanks for tackling this feature request! It would be nice to finally have SCTP support :smiley: 

My main concern with the current changes is the many limitations. They need to be documented, but I think it would also be best to put the feature behind a flag rather than introduce many runtime checks. (I'm actually surprised you didn't hit any complexity issues because of that.)

The commit history also needs some work. Each commit should contain changes that logically belong together, in a way that facilitates review of the different aspects of this new feature.

Commit:
policy: Add support for SCTP policies.

This adds support L4 SCTP policies. Note that L7 policies on top of SCTP
are unsupported.

One behavior change is that policymap will no longer panic on unknown
protos. Previously, it used to explicitly ignore UDP and panic on
unknown protocols. Now it ignores all unknown protocols.

Signed-off-by: Harsh Modi <harshmodi@google.com>

Commit:
hubble: Add SCTP support.

Allow tracing SCTP packets.

Signed-off-by: Harsh Modi <harshmodi@google.com>

Commit:
bpf: Add partial SCTP support.

Signed-off-by: Harsh Modi <harshmodi@google.com>
Co-authored-by: Nathan Sweet <nathanjsweet@users.noreply.github.com>

Commit:
docs: Add basic docs on SCTP.

Add docs on how to enable and the limitations of SCTP support.

Signed-off-by: Harsh Modi <harshmodi@google.com>

Commit:
test: Add a BPF test to exercise SCTP.

This test is a clone of hairpin_flow, but for SCTP packets.

Signed-off-by: Harsh Modi <harshmodi@google.com>

qmonnet commented on test/k8s/sctp.go:
```diff
@@ -0,0 +1,127 @@
+		kubectl.CloseSSHClient()
+	})
+
+	It("Tests SCTP", func() {
```
> So my awesome SCTP test depends on `CONFIG_IP_SCTP`, which COS does not enable :) What is the preferred approach for selectively disabling this test?

This should skip the test on GKE:


```suggestion
	SkipItIf(helpers.RunsOnGKE, "Tests SCTP", func() {
```

DolceTriade replied:
Thanks!

joestringer commented on api/v1/openapi.yaml:
```diff
@@ -3,7 +3,7 @@ swagger: '2.0'
   title: Cilium API
   description: Cilium
-  version: v1beta
+  version: v1beta1
```
Why is it necessary to change the version number? I think this is probably OK, but it could be helpful to have this in a separate dedicated commit with the rationale, that way we can try to ensure that any future changes to the version string can match whatever thing you had trouble with here.

DolceTriade replied:
Since there was a change to the API (the addition of the SCTP stuff), I thought it would make sense to bump it. If that isn't the protocol, we can revert this change.

DolceTriade replied:
Do you still want this in a separate commit than the one that changes the API?

joestringer replied:
Overall on this unix-socket API, there is no intention to make it a public, stable API. The stable APIs are typically the ones that interface with Kubernetes, structures we store in key-value stores, or the Hubble flow APIs. Generally we won't break things on this API if there's a reasonable solution to change something without breaking it, but for instance in this case it seems like there's just new numbers & fields that should be backwards-compatible anyway.

If CI's happy with this then I'm OK with it. I see it's in the same commit as the other API changes anyway so that'll do for now.

joestringer commented on bpf/Makefile:
```diff
@@ -78,7 +78,7 @@ LB_OPTIONS = \
 	-DENABLE_IPV6:-DENCAP_IFINDEX:-DTUNNEL_MODE:-DENABLE_IPSEC:-DIP_POOLS:-DENABLE_HOST_SERVICES_UDP:-DENABLE_HOST_SERVICES_TCP:-DENABLE_NODEPORT:-DENABLE_NODEPORT_ACCELERATION:-DENABLE_SESSION_AFFINITY:-DENABLE_BANDWIDTH_MANAGER:-DENABLE_SRC_RANGE_CHECK:-DLB_SELECTION:-DLB_SELECTION_MAGLEV: \
 	-DENABLE_IPV6:-DENCAP_IFINDEX:-DTUNNEL_MODE:-DENABLE_IPSEC:-DIP_POOLS:-DENABLE_HOST_SERVICES_UDP:-DENABLE_HOST_SERVICES_TCP:-DENABLE_NODEPORT:-DENABLE_NODEPORT_ACCELERATION:-DENABLE_SESSION_AFFINITY:-DENABLE_BANDWIDTH_MANAGER:-DENABLE_SRC_RANGE_CHECK:-DLB_SELECTION:-DLB_SELECTION_MAGLEV:-DENABLE_SOCKET_LB_HOST_ONLY: \
-	-DENABLE_IPV6:-DENCAP_IFINDEX:-DTUNNEL_MODE:-DENABLE_IPSEC:-DIP_POOLS:-DENABLE_HOST_SERVICES_UDP:-DENABLE_HOST_SERVICES_TCP:-DENABLE_NODEPORT:-DENABLE_NODEPORT_ACCELERATION:-DENABLE_SESSION_AFFINITY:-DENABLE_BANDWIDTH_MANAGER:-DENABLE_SRC_RANGE_CHECK:-DLB_SELECTION:-DLB_SELECTION_MAGLEV:-DENABLE_SOCKET_LB_HOST_ONLY:-DENABLE_L7_LB:
+	-DENABLE_IPV6:-DENCAP_IFINDEX:-DTUNNEL_MODE:-DENABLE_IPSEC:-DIP_POOLS:-DENABLE_HOST_SERVICES_UDP:-DENABLE_HOST_SERVICES_TCP:-DENABLE_NODEPORT:-DENABLE_NODEPORT_ACCELERATION:-DENABLE_SESSION_AFFINITY:-DENABLE_BANDWIDTH_MANAGER:-DENABLE_SRC_RANGE_CHECK:-DLB_SELECTION:-DLB_SELECTION_MAGLEV:-DENABLE_SOCKET_LB_HOST_ONLY:-DENABLE_L7_LB
```
Why does this change remove the trailing colon? It seems like you intended instead to have the line finish with `-DENABLE_SCTP:`.

DolceTriade replied:
Oops. I originally added ENABLE_SCTP there, but moved it elsewhere after.

joestringer commented on Documentation/configuration/sctp.rst:
```diff
@@ -0,0 +1,28 @@
+
+[*] SCTP support does not support rewriting ports for SCTP packets. This means
+that when defining services, the targetPort **MUST** equal the port, otherwise
+the packet will be dropped.
```
In @pchaigno's comment [here](https://github.com/cilium/cilium/pull/20033/files#r895211322), there is a longer list of current limitations of SCTP support with various Cilium features. Does it make sense to list each of those here to make it clear what kind of incompatibilities exist?

DolceTriade replied:
I'll list them here to be explicit, since the downstream effects of not being able to change ports may not be immediately clear.

joestringer commented on bpf/lib/l4.h:
```diff
@@ -37,7 +37,8 @@ static __always_inline int l4_modify_port(struct __ctx_buff *ctx, int l4_off,
 {
-	if (csum_l4_replace(ctx, l4_off, csum_off, old_port, port, sizeof(port)) < 0)
+	if (csum_off->offset > 0 &&
+	    csum_l4_replace(ctx, l4_off, csum_off, old_port, port, sizeof(port)) < 0)
```
If the caller is already dropping SCTP before it hits this line (I'm looking at `reverse_map_l4_port()`), is this change still necessary?

DolceTriade replied:
No, SCTP packets are dropped before entering all instances of this function. Probably a remnant of before we added explicit drop for SCTP packets that need to change ports.

joestringer commented on bpf/lib/lb.h:
```diff
@@ -396,8 +402,14 @@ static __always_inline int reverse_map_l4_port(struct __ctx_buff *ctx, __u8 next
+				/* This will change the SCTP checksum, which we cannot fix right now.
+				 * This will likely need kernel changes before we can drop this.
+				 */
+				if (nexthdr == IPPROTO_SCTP) return DROP_CSUM_L4;
```
Please put the `return DROP_CSUM_L4` on a newline.

DolceTriade replied:
Done.

joestringer commented on Documentation/configuration/sctp.rst:
```diff
@@ -0,0 +1,28 @@
+Pass ``--set sctp.enabled=true`` to helm.
+
+Limitations
+===================
```
Nit, please follow the standard style for headings:

```suggestion
Enabling
========

Pass ``--set sctp.enabled=true`` to helm.

Limitations
===========

```

DolceTriade replied:
Done.

joestringer commented on Documentation/configuration/sctp.rst:
```diff
@@ -0,0 +1,28 @@
+Limitations
+===================
+Cilium supports basic SCTP support. Specifically, the following is supported:
+ - Pod <-> Pod communication
```
```suggestion
 - Pod <-> Pod communication
 - Pod <-> Pod communication with network policies applied to SCTP traffic
```

joestringer reviewed (changes requested):
I took a brief overview of the feature overall, seems like there's still one or two missing bits of docs around the limitations. I noticed some style nits in the bpf and docs code as well.

There's currently some ongoing work around how Cilium does e2e testing, so it would be nice to get some feedback from folks actively working on the next generation of CI to see how that portion of the PR can be made amenable to the upcoming changes.

gandro commented on api/v1/flow/flow.proto:
```diff
@@ -126,6 +126,7 @@ message Layer4 {
     oneof protocol {
         TCP TCP = 1;
         UDP UDP = 2;
+        SCTP SCTP = 5;
```
optional nit: I prefer this at the bottom, so the order remains consistent with the assigned number.

DolceTriade replied:
Oops, missed this comment. Done.

gandro commented on pkg/hubble/parser/seven/parser.go:
```diff
@@ -233,6 +233,15 @@ func decodeLayer4(protocol accesslog.TransportProtocol, source, destination acce
 				},
 			},
 		}, uint16(source.Port), uint16(destination.Port)
+	case u8proto.SCTP:
```
Not sure this case is needed, does any of our L7 proxies (dns or envoy) support SCTP? If not, then this will never be hit.
On the other hand, it also doesn't hurt if we ever add L7 support for an SCTP based protocol.

DolceTriade replied:
Seems better to be thorough here if it is all the same. 🤷

gandro reviewed (approved):
Helm, Hubble and Monitor CODEOWNERs look good to me.

Commit:
api: Add support for SCTP.

Allow specifying SCTP as a protocol for policy tracing.

Signed-off-by: Harsh Modi <harshmodi@google.com>

pchaigno commented on proxylib/proxylib/policymap.go:
```diff
@@ -182,10 +182,6 @@ func newPortNetworkPolicies(config []*cilium.PortNetworkPolicy, dir string) Port
 
-		if rule.GetProtocol() != core.SocketAddress_TCP {
-			ParseError(fmt.Sprintf("Invalid transport protocol %v", rule.GetProtocol()), config)
-		}
```
The behavior is changed here: we used to panic if the protocol was unknown, now we simply skip it.

DolceTriade replied:
Yes, I kept failing the connectivity test if I didn't do this. But good note. I'll call this out in my commit message.

DolceTriade replied:
Before, it used to explicitly ignore UDP. Would it be better to also explicitly ignore SCTP as well?

pchaigno replied:
I think that would be best, yes, to preserve the previous behavior.

DolceTriade replied:
Reverted this change. For some reason I can no longer reproduce the failure. Maybe it will show up in the presubmits 🤷

DolceTriade replied:
Furthermore, There is no way to explicitly skip SCTP because it simply doesn't exist as an option in the enum, so it's either no change at all or ignore all non-TCP protocols.

pchaigno commented on bpf/Makefile:
```diff
@@ -69,7 +69,7 @@ LB_OPTIONS = \
 	-DENABLE_IPV4:-DENABLE_IPV6:-DENCAP_IFINDEX:-DTUNNEL_MODE:-DENABLE_IPSEC:-DIP_POOLS:-DENABLE_HOST_SERVICES_UDP:-DENABLE_HOST_SERVICES_TCP:-DENABLE_NODEPORT:-DENABLE_NODEPORT_ACCELERATION: \
 	-DENABLE_IPV4:-DENABLE_IPV6:-DENCAP_IFINDEX:-DTUNNEL_MODE:-DENABLE_IPSEC:-DIP_POOLS:-DENABLE_HOST_SERVICES_UDP:-DENABLE_HOST_SERVICES_TCP:-DENABLE_NODEPORT:-DENABLE_NODEPORT_ACCELERATION:-DENABLE_SESSION_AFFINITY: \
-	-DENABLE_IPV4:-DENABLE_IPV6:-DENCAP_IFINDEX:-DTUNNEL_MODE:-DENABLE_IPSEC:-DIP_POOLS:-DENABLE_HOST_SERVICES_UDP:-DENABLE_HOST_SERVICES_TCP:-DENABLE_NODEPORT:-DENABLE_NODEPORT_ACCELERATION:-DENABLE_SESSION_AFFINITY:-DENABLE_L7_LB: \
+	-DENABLE_IPV4:-DENABLE_IPV6:-DENCAP_IFINDEX:-DTUNNEL_MODE:-DENABLE_IPSEC:-DIP_POOLS:-DENABLE_HOST_SERVICES_UDP:-DENABLE_HOST_SERVICES_TCP:-DENABLE_NODEPORT:-DENABLE_NODEPORT_ACCELERATION:-DENABLE_SESSION_AFFINITY:-DENABLE_L7_LB:-DENABLE_SCTP: \
```
`ENABLE_SCTP` is used in all BPF programs and not just bpf_sock/bpf_overlay, no? So we should also add it to the other entries for e.g. bpf_host and bpf_lxc.

pchaigno replied:
We also need to add `ENABLE_SCTP` to the complexity tests in `bpf/complexity-tests/`.

DolceTriade replied:
Ah I didn't realize that adding it here didn't add it everywhere.

pchaigno replied:
Yeah, it's a bit more complicated than it should be right now. We'll likely generate the complexity-tests files in the long term but for now we need to modify them manually.

DolceTriade replied:
Done.

pchaigno commented on bpf/bpf_lxc.c:
```diff
@@ -56,7 +56,8 @@
     defined(ENABLE_SOCKET_LB_HOST_ONLY) || \
-    defined(ENABLE_L7_LB)
+    defined(ENABLE_L7_LB)               || \
+    defined(ENABLE_SCTP)
```
What's the rationale for this change? AFAIU, since we can't change the ports, neither socket-level LB nor per-packet LB will be supported anyway.

DolceTriade replied:
Hm, I thought that per-packet LB was required to do lookups for service IPs. I recall having weird issues without this enabled (something where we would continue to send packets with the Service IP instead of the translated Pod IP). I can try again with this disabled and see what happens.

pchaigno replied:
Oh, right! Socket-level LB is the one case of DNAT where things will work fine despite not being able to change the checksum, because we DNAT before the packets are even generated. So this change makes sense. Maybe the documentation should clarify this point though.

pchaigno commented on bpf/lib/lb.h:
```diff
@@ -396,8 +402,15 @@ static __always_inline int reverse_map_l4_port(struct __ctx_buff *ctx, __u8 next
+#endif  /* ENABLE_SCTP */
 				ret = l4_modify_port(ctx, l4_off, TCP_SPORT_OFF,
-						     csum_off, port, old_port);
+							csum_off, port, old_port);
```
Seems this was changed by mistake?

DolceTriade replied:
Yeah, indenting with tabs and spaces is always weird for me :(

pchaigno replied:
Indenting with tabs, aligning with spaces, *à la* Linux :slightly_smiling_face:

pchaigno commented on bpf/lib/lb.h:
```diff
@@ -674,10 +687,20 @@ static __always_inline int lb6_xlate(struct __ctx_buff *ctx,
+		 */
+		if (nexthdr == IPPROTO_SCTP)
+			return DROP_CSUM_L4;
+#endif  /* ENABLE_SCTP */
```
We could check for this right after `l4_xlate:` to avoid having to modify the existing condition.

DolceTriade replied:
I figured this would have to change after port translation support is added, but I can change it if you prefer.

pchaigno replied:
Yep, it doesn't sound like port translation is right around the corner (given it requires kernel changes), so best to improve now I think.

DolceTriade replied:
Ack. Done.

pchaigno commented on bpf/lib/lb.h:
```diff
@@ -1314,11 +1340,21 @@ lb4_xlate(struct __ctx_buff *ctx, __be32 *new_daddr, __be32 *new_saddr __maybe_u
+		 */
+		if (nexthdr == IPPROTO_SCTP)
+			return DROP_CSUM_L4;
+#endif  /* ENABLE_SCTP */
```
Ditto.

DolceTriade replied:
Done.

pchaigno commented on Documentation/configuration/sctp.rst:
```diff
@@ -0,0 +1,33 @@
+*******************
+SCTP support (BETA)
+*******************
+
```
Please add:
```
.. include:: ../beta.rst
```
here. See other guides for beta features doing the same.

DolceTriade replied:
Done

pchaigno commented on Documentation/configuration/sctp.rst:
```diff
@@ -0,0 +1,33 @@
+.. _sctp:
+
+*******************
+SCTP support (BETA)
```
```suggestion
SCTP support (beta)
```

DolceTriade replied:
Done.

pchaigno commented on pkg/option/config.go:
```diff
@@ -663,6 +663,9 @@ const (
 	EnableIPv6NDPName = "enable-ipv6-ndp"
 
+	// EnableSCTPName is the name of the option to enable SCTP support
+	EnableSCTPName = "enable-sctp"
```
Don't we also need to add this flag in `initializeFlags`? I'm surprised it even works without that change.

DolceTriade replied:
Hm, interesting that it works without that. Added.

Commit:
helm: Add sctp.enabled and use it to set ENABLE_SCTP.

Signed-off-by: Harsh Modi <harshmodi@google.com>

joestringer commented on bpf/lib/lb.h:
```diff
@@ -396,8 +402,15 @@ static __always_inline int reverse_map_l4_port(struct __ctx_buff *ctx, __u8 next
+#endif  /* ENABLE_SCTP */
 				ret = l4_modify_port(ctx, l4_off, TCP_SPORT_OFF,
-						     csum_off, port, old_port);
+									 csum_off, port, old_port);
```
nit: Spacing.

DolceTriade replied:
Hm, make checkpatch doesn't seem to catch this...and that is pretty strict about aligning stuff. Maybe it's fixed now...

joestringer replied:
Did you push your updated branch?

For what it's worth, kernel style expects that tabs are the width of 8 spaces, so however you configure your editor's tabs for editing C files, that is worth considering.

joestringer commented on test/k8s/manifests/sctp/iperf3.yaml:
```diff
@@ -0,0 +1,101 @@
+          effect: NoSchedule
+      containers:
+      - name: iperf3-server
+        image: networkstatic/iperf3
```
It looks like the full image path for this is `docker.io/networkstatic/iperf3`, which unfortunately is on dockerhub, hence can hit issues with dockerhub ratelimiting.

This (and other image references) should at least contain the full path + version string + sha of the image, so that we can manage any potential (breaking) version changes from the upstream docker image maintainer.

Furthermore, we should pre-pull the image into the VM image as well to avoid issues where the CI fails because of dockerhub ratelimiting. For that, it needs to be integrated here: https://github.com/cilium/packer-ci-build/blob/master/provision/pull-images.sh . Maybe we can work together on Slack to get the VM image bits updated correctly to avoid this issue.

DolceTriade replied:
I see. I asked about this on Slack and was told it was fine: https://cilium.slack.com/archives/C7PE7V806/p1655397271386549.

It's not a super complicated image (just ubuntu + iperf3), so we can create it anywhere. I'll sync up with you tomorrow then to find a path forward. Thanks for walking me through all of this! 😄

joestringer replied:
:+1: yeah I think the image itself should be fine, there's just two remaining nits: (1), the name should be fully specified so that we don't randomly break CI because upstream decides to push a new image that's somehow incompatible/broken/etc and (2) the image should be cached so that we don't break CI because of dockerhub ratelimiting.

DolceTriade replied:
https://github.com/cilium/packer-ci-build/pull/339

DolceTriade replied:
Also updated the test to use a fully qualified name with hash.

joestringer replied:
Now merged: #20596

joestringer reviewed (commented):
Looks good overall, couple of remaining minor nits from my side. I suspect that we'll need to resolve the image pull question before merging in order to prevent this test from causing CI failures due to dockerhub ratelimiting.

joestringer reviewed (commented):
Just one remaining comment - if we're specifically avoiding running this on Linux 4.9 due to complexity issues or something of that nature, then it would be nice to have some extra ProbeManager check in the daemon to tell the user that they shouldn't enable this option if their kernel is too old. We could maybe switch that on `HaveLargeInsnLimit`?

cc also @pchaigno 

Other than that, the PR LGTM.

Commit:
sctp: Error out if SCTP is enabled on older kernels.

SCTP requires many more branches and paths which will cause the programs
to exceed the instruction limit on older kernels.

Signed-off-by: Harsh Modi <harshmodi@google.com>

nathanjsweet commented on bpf/lib/lb.h:
```diff
@@ -396,8 +402,15 @@ static __always_inline int reverse_map_l4_port(struct __ctx_buff *ctx, __u8 next
 			if (port != old_port) {
+#ifdef ENABLE_SCTP
+				/* This will change the SCTP checksum, which we cannot fix right now.
+				 * This will likely need kernel changes before we can drop this.
```
```suggestion
				 * This will likely need kernel changes before we can remove this.
```
Did you mean remove this code? The word "drop" is confusing in this context.

nathanjsweet commented on bpf/lib/lb.h:
```diff
@@ -685,6 +698,14 @@ static __always_inline int lb6_xlate(struct __ctx_buff *ctx,
 l4_xlate:
+#ifdef ENABLE_SCTP
+	/* This will change the SCTP checksum, which we cannot fix right now.
+	 * This will likely need kernel changes before we can drop this.
```
```suggestion
	 * This will likely need kernel changes before we can remove this.
```

nathanjsweet commented on bpf/lib/lb.h:
```diff
@@ -1340,12 +1364,20 @@ lb4_xlate(struct __ctx_buff *ctx, __be32 *new_daddr, __be32 *new_saddr __maybe_u
 l4_xlate:
+#ifdef ENABLE_SCTP
+	/* This will change the SCTP checksum, which we cannot fix right now.
+	 * This will likely need kernel changes before we can drop this.
```
```suggestion
	 * This will likely need kernel changes before we can remove this.
```

nathanjsweet reviewed (changes requested):
I represent approval for the API changes. They look good. A couple of nits on your comments.

joamaki reviewed (approved):
The bpf test case looks good, thanks!

pchaigno commented on bpf/Makefile:
```diff
@@ -215,7 +215,7 @@ XDP_OPTIONS = $(LB_OPTIONS) \
 	-DENABLE_NODEPORT_ACCELERATION:-DENABLE_IPV4:-DENABLE_IPV6:-DENABLE_NODEPORT:-DENABLE_DSR:-DDSR_ENCAP_MODE:-DDSR_ENCAP_IPIP:-DDSR_ENCAP_NONE=2 \
 	-DENABLE_NODEPORT_ACCELERATION:-DENABLE_IPV4:-DENABLE_IPV6:-DENABLE_NODEPORT:-DENABLE_DSR:-DENABLE_CAPTURE:-DDSR_ENCAP_MODE:-DDSR_ENCAP_NONE:-DDSR_ENCAP_IPIP=2 \
-	-DENABLE_NODEPORT_ACCELERATION:-DENABLE_IPV4:-DENABLE_IPV6:-DENABLE_NODEPORT:-DENABLE_DSR:-DENABLE_CAPTURE:-DDSR_ENCAP_MODE:-DDSR_ENCAP_IPIP:-DDSR_ENCAP_NONE=2
+	-DENABLE_NODEPORT_ACCELERATION:-DENABLE_IPV4:-DENABLE_IPV6:-DENABLE_NODEPORT:-DENABLE_DSR:-DENABLE_CAPTURE:-DDSR_ENCAP_MODE:-DDSR_ENCAP_IPIP:-DDSR_ENCAP_NONE=2:-DENABLE_SCTP
```
This is a bit tricky, but any `:` will be replaced by `=1`, so whenever we set our own value (`=2` here) we need to omit the `:`. That requires us to have macros with a value at the end of the line. Very hackish but I couldn't find a better way.

```suggestion
	-DENABLE_NODEPORT_ACCELERATION:-DENABLE_IPV4:-DENABLE_IPV6:-DENABLE_NODEPORT:-DENABLE_DSR:-DENABLE_CAPTURE:-DDSR_ENCAP_MODE:-DDSR_ENCAP_IPIP:-DENABLE_SCTP:-DDSR_ENCAP_NONE=2
```

DolceTriade replied:
Done.

pchaigno commented on bpf/complexity-tests/49/bpf_overlay.txt:
```diff
@@ -1 +1 @@
--DSKIP_DEBUG=1 -DENABLE_IPV4=1 -DENABLE_IPV6=1 -DENABLE_SOCKET_LB_TCP=1 -DENABLE_SOCKET_LB_UDP=1 -DENABLE_ROUTING=1 -DNO_REDIRECT=1 -DPOLICY_VERDICT_NOTIFY=1 -DALLOW_ICMP_FRAG_NEEDED=1 -DENABLE_IDENTITY_MARK=1 -DMONITOR_AGGREGATION=3 -DCT_REPORT_FLAGS=0x0002 -DENABLE_HOST_FIREWALL=1 -DENABLE_ICMP_RULE=1 -DENABLE_CUSTOM_CALLS=1 -DENABLE_IPSEC=1 -DIP_POOLS=1 -DENCAP_IFINDEX=1 -DTUNNEL_MODE=1
\ No newline at end of file
+-DSKIP_DEBUG=1 -DENABLE_IPV4=1 -DENABLE_IPV6=1 -DENABLE_SOCKET_LB_TCP=1 -DENABLE_SOCKET_LB_UDP=1 -DENABLE_ROUTING=1 -DNO_REDIRECT=1 -DPOLICY_VERDICT_NOTIFY=1 -DALLOW_ICMP_FRAG_NEEDED=1 -DENABLE_IDENTITY_MARK=1 -DMONITOR_AGGREGATION=3 -DCT_REPORT_FLAGS=0x0002 -DENABLE_HOST_FIREWALL=1 -DENABLE_ICMP_RULE=1 -DENABLE_CUSTOM_CALLS=1 -DENABLE_IPSEC=1 -DIP_POOLS=1 -DENCAP_IFINDEX=1 -DTUNNEL_MODE=1
```
Why was this file touched?

DolceTriade replied:
Reverted

pchaigno replied:
It seems there's now a newline at the end of the file.

pchaigno commented on bpf/complexity-tests/49/bpf_xdp.txt:
```diff
@@ -1 +1 @@
--DSKIP_DEBUG=1 -DENABLE_IPV4=1 -DENABLE_IPV6=1 -DENABLE_SOCKET_LB_TCP=1 -DENABLE_SOCKET_LB_UDP=1 -DENABLE_ROUTING=1 -DNO_REDIRECT=1 -DPOLICY_VERDICT_NOTIFY=1 -DALLOW_ICMP_FRAG_NEEDED=1 -DENABLE_IDENTITY_MARK=1 -DMONITOR_AGGREGATION=3 -DCT_REPORT_FLAGS=0x0002 -DENABLE_HOST_FIREWALL=1 -DENABLE_ICMP_RULE=1 -DENABLE_CUSTOM_CALLS=1 -DENABLE_IPSEC=1 -DIP_POOLS=1 -DENABLE_PREFILTER=1
\ No newline at end of file
+-DSKIP_DEBUG=1 -DENABLE_IPV4=1 -DENABLE_IPV6=1 -DENABLE_SOCKET_LB_TCP=1 -DENABLE_SOCKET_LB_UDP=1 -DENABLE_ROUTING=1 -DNO_REDIRECT=1 -DPOLICY_VERDICT_NOTIFY=1 -DALLOW_ICMP_FRAG_NEEDED=1 -DENABLE_IDENTITY_MARK=1 -DMONITOR_AGGREGATION=3 -DCT_REPORT_FLAGS=0x0002 -DENABLE_HOST_FIREWALL=1 -DENABLE_ICMP_RULE=1 -DENABLE_CUSTOM_CALLS=1 -DENABLE_IPSEC=1 -DIP_POOLS=1 -DENABLE_PREFILTER=1
```
Ditto.

DolceTriade replied:
Reverted.

pchaigno replied:
Ditto.

pchaigno commented on bpf/lib/lb.h:
```diff
@@ -396,8 +402,15 @@ static __always_inline int reverse_map_l4_port(struct __ctx_buff *ctx, __u8 next
+#endif  /* ENABLE_SCTP */
 				ret = l4_modify_port(ctx, l4_off, TCP_SPORT_OFF,
-						     csum_off, port, old_port);
+									 csum_off, port, old_port);
```
There's an indentation issue here.

DolceTriade replied:
I keep forgetting kernel indent width is 8 :(

pchaigno commented on daemon/cmd/daemon.go:
```diff
@@ -997,6 +997,12 @@ func NewDaemon(ctx context.Context, cleaner *daemonCleanup, epMgr *endpointmanag
 	}
+	if option.Config.EnableSCTP {
+		if probes.HaveLargeInstructionLimit() != nil {
+			log.WithError(err).Error("SCTP needs support kernel 5.2 or newer")
```
```suggestion
			log.WithError(err).Error("SCTP support needs kernel 5.2 or newer")
```

DolceTriade replied:
done

pchaigno reviewed (changes requested):
A couple changes needed below but they should all be trivial to implement.

pchaigno reviewed (changes requested):
I think there was a small mishaps while addressing the changes above.

gandro reviewed (approved):
FWIW, Hubble still looks good
